File: src/exemplos/05_EstruturasDeDados/06_Arquivo/04-texto.c
/**      @file: 04-texto.c
 *     @author: Guilherme N. Ramos (gnramos@unb.br)
 * @disciplina: Algoritmos e Programação de Computadores
 *
 * Exemplo de leitura de um arquivo texto e escrita em outro
 * arquivo. */

#include <stdio.h>
#include <stdlib.h>

int copia(char* arquivo) {
  char backup[500], string[500];

  FILE* saida = NULL;
  FILE* entrada = fopen(arquivo, "r");

  if(entrada == NULL) {
    printf("Não foi possível abrir \"%s\".\n", arquivo);
    return EXIT_FAILURE;
  }

  /* Definição do identificador do novo arquivo. */
  sprintf(backup, "copia_de_%s", arquivo);
  saida = fopen(backup, "w+");

  if(saida == NULL) {
    printf("\nNão foi possível abrir \"%s\".\n", backup);
    fclose(entrada); /* <-  */
    return EXIT_FAILURE;
  }

  /* Escreve cabeçalho. */
  fprintf(saida, "/**      @file: %s\n", backup);
  fprintf(saida, " *     @author: [gerado automaticamente]\n");
  fprintf(saida, " * @disciplina: Algoritmos e Programação de Computadores\n");
  fprintf(saida, " *\n * Exemplo de manipulação de arquivos. Este arquivo foi\n");
  fprintf(saida, " * gerado a partir de \"%s\". */\n\n", arquivo);

  /* Copia linha a linha. */
  while(NULL != fgets(string, sizeof(string), entrada))
    fputs(string, saida);

  fclose(saida);
  fclose(entrada);

  return EXIT_SUCCESS;
 }


int main(int argc, char** argv) {
  if(argc < 2) {
    printf("É preciso fornecer um arquivo texto como argumento.\n");
    return EXIT_FAILURE;
  }

  if(EXIT_SUCCESS != copia(argv[1])) {
    printf("Houve um erro ao tentar copiar \"%s\"\n", argv[1]);
  }

  return EXIT_SUCCESS;
}

File: src/exemplos/05_EstruturasDeDados/06_Arquivo/04-texto.py
#  -*- coding: utf-8 -*-
#    @package: 03-texto.py
#     @author: Guilherme N. Ramos (gnramos@unb.br)
# @disciplina: Algoritmos e Programação de Computadores
#
# Exemplo de leitura de um arquivo texto e escrita em outro
# arquivo.


import sys


def copia(arquivo):
    '''Cria um arquivo como cópia do arquivo de nome dado como argumento.
    Retorna True caso seja possível, False caso contrário.
    '''
    try:
        with open(arquivo, 'r') as f:
            conteudo_original = f.read()

        backup = 'copia_de_' + arquivo
        with open(backup, 'w') as f:
            f.write('#  -*- coding: utf-8 -*-\n'
                    '#    @package: ' + backup + '\n'
                    '#     @author: [gerado automaticamente]\n'
                    '# @disciplina: Algoritmos e Programação de Computadores\n'
                    '#\n'
                    '# Exemplo de manipulação de arquivos. Este arquivo foi\n'
                    '# gerado a partir de \'' + arquivo + '\'.\n\n')
            f.write(conteudo_original)
        return True
    except:
        print('Erro!')
        return False


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('É preciso fornecer um arquivo texto como argumento.')
        exit(1)

    if not copia(sys.argv[1]):
        print('Houve um erro ao tentar copiar \'%s\'' % sys.argv[1])
